packages/doenetml-worker-javascript/src/utils/dast/types.ts

```diff
@@ -0,0 +1,243 @@
+    component: SerializedComponent;
+    ignoreFixed?: boolean;
+};
+
+function isComponentAttribute(obj: unknown): obj is ComponentAttribute {
+    const typedObj = obj as ComponentAttribute;
+    return (
+        typeof typedObj === "object" &&
+        typedObj !== null &&
+        typedObj.type === "component" &&
+        typeof typedObj.name === "string" &&
+        isSerializedComponent(typedObj.component) &&
+        (typedObj.ignoreFixed === undefined ||
+            typeof typedObj.ignoreFixed === "boolean")
+    );
+}
+
+/** An attribute whose value is a primitive data type */
+type PrimitiveAttribute = {
+    type: "primitive";
+    name: string;
+    primitive: PrimitiveAttributeValue;
+    ignoreFixed?: boolean;
+};
+
+function isPrimitiveAttribute(obj: unknown): obj is PrimitiveAttribute {
+    const typedObj = obj as PrimitiveAttribute;
+    return (
+        typeof typedObj === "object" &&
+        typedObj !== null &&
+        typedObj.type === "primitive" &&
+        typeof typedObj.name === "string" &&
+        isPrimitiveAttributeValue(typedObj.primitive) &&
+        (typedObj.ignoreFixed === undefined ||
+            typeof typedObj.ignoreFixed === "boolean")
+    );
+}
+
+/**
+ * An attribute that has not yet been resolved from determining its parent component type.
+ * Typically used to store attributes in composites before they are added to its replacements.
+ */
+type UnresolvedAttribute = {
+    type: "unresolved";
+    name: string;
+    childrenForFutureComponent: (SerializedComponent | string)[];
+    ignoreFixed?: boolean;
+};
+
+function isUnresolvedAttribute(obj: unknown): obj is UnresolvedAttribute {
+    const typedObj = obj as UnresolvedAttribute;
+    return (
+        typeof typedObj === "object" &&
+        typedObj !== null &&
+        typedObj.type === "unresolved" &&
+        typeof typedObj.name === "string" &&
+        Array.isArray(typedObj.childrenForFutureComponent) &&
+        typedObj.childrenForFutureComponent.every(
+            (child) =>
+                typeof child === "string" || isSerializedComponent(child),
+        ) &&
+        (typedObj.ignoreFixed === undefined ||
+            typeof typedObj.ignoreFixed === "boolean")
+    );
+}
+
+/** An instance of an attribute in a `SerializedComponent` */
+export type SerializedAttribute =
+    | ComponentAttribute
+    | PrimitiveAttribute
+    | UnresolvedAttribute;
+
+export function isSerializedAttribute(
+    obj: unknown,
+): obj is SerializedAttribute {
+    return (
+        isComponentAttribute(obj) ||
+        isPrimitiveAttribute(obj) ||
+        isUnresolvedAttribute(obj)
+    );
+}
+
+export type ErrorRecord = {
+    message: string;
+    position?: Position;
+};
+
+export type WarningRecord = ErrorRecord & { level: number };
+
+/**
+ * Information about the definition of an attribute for a component type.
+ *
+ * Typically most fields are undefined, as many cover specialized cases.
+ */
+export type AttributeDefinition<T> = {
```

Comment:
So nice to have these listed!

Reply:
Yes! Since the code I extracted these from has not been converted to TypeScript, I'm not confident I got every field. But most of them.